### PR TITLE
Support arbitrary statistic interval for flow control && stat parameters configurable

### LINKED ISSUE #129: Enable configuration for sample count & interval of leap array
## Issue Description

Type: *feature request*

### Describe what feature you want

Enable configuration for sample count & interval of leap array.

1. make the global const DefaultSampleCountTotal & DefaultIntervalMsTotal configurable;
2. make SampleCount & IntervalMs per Resource configurable.

### Additional context

### PR DESCRIPTION
**Pull request by louyuting**
### Describe what this PR does / why we need it
Support arbitrary statistic interval for normal flow rule
stat parameters configurable

### Does this pull request fix one issue?
Resolve 
#129 
#167 

### Describe how you did it
Support arbitrary statistic interval for normal flow rule

1. If user doesn't set StatisticConfig, that means using default metric statistic.
2.If user specifies StatisticConfig and meets the compliance of CheckValidityForReuseStatistic,
    will build readonly metric statistic based on resource's global statistic.
 3.If user specifies StatisticConfig and doesn't meet the compliance of CheckValidityForReuseStatistic,
    will generate independent token bucket statistic for this rule.

### Describe how to verify it
TODO
1. unit test
2. integration test

### Special notes for reviews

### CODE REVIEW

**Comment by louyuting** [comment]
@sczyh30  Could you please review this large PR?

File: core/flow/rule_manager.go [hunk 1]
```diff
@@ -7,44 +7,92 @@ import (
...
+		return tsc, nil
 	}
 }
 
 func logRuleUpdate(m TrafficControllerMap) {
 	bs, err := json.Marshal(rulesFrom(m))
 	if err != nil {
-		logging.Info("[FlowRuleManager] Flow rules loaded")
+		if len(m) == 0 {
+			logging.Info("[FlowRuleManager] Flow rules was cleared")
```
Comment (yunfeiyanggzq) [inline core/flow/rule_manager.go:82]: `was` ->`were`

File: core/flow/rule_manager.go [hunk 2]
```diff
@@ -7,44 +7,92 @@ import (
...
+		} else {
+			logging.Info("[FlowRuleManager] Flow rules was loaded")
```
Comment (yunfeiyanggzq) [inline core/flow/rule_manager.go:84]: same

File: core/flow/rule_manager.go [hunk 3]
```diff
@@ -73,26 +121,59 @@ func onRuleUpdate(rules []*FlowRule) (err error) {
...
+func GetRules() []Rule {
+	rules := getRules()
+	ret := make([]Rule, 0, len(rules))
+	for _, rule := range rules {
+		ret = append(ret, *rule)
+	}
+	return ret
+}
+
+func GetResRules(res string) []Rule {
```
Comment (yunfeiyanggzq) [inline core/flow/rule_manager.go:161]: maybe the name can be improved

File: core/flow/rule_manager.go [hunk 4]
```diff
@@ -207,24 +294,40 @@ func IsValidFlowRule(rule *FlowRule) error {
...
-	return checkControlBehaviorField(rule)
+	if err := checkStatisticConfig(rule); err != nil {
+		return err
+	}
+
+	return checkControlStrategyField(rule)
 }
 
-func checkClusterField(rule *FlowRule) error {
+func checkClusterField(rule *Rule) error {
```
Comment (yunfeiyanggzq) [inline core/flow/rule_manager.go:304]: I think we'd better judge the two parameters separately for more cluster parameters in the future

File: core/base/stat.go [hunk 5]
```diff
@@ -46,4 +50,47 @@ type StatNode interface {
...
+	}
+	return nil
+}
+
+// CheckValidityForReuseStatistic check the compliance whether readonly metric statistic can be built based on resource's global statistic
+// The parameters, sampleCount and intervalInMs, are the parameters of the metric statistic you want to build
+// The parameters, parentSampleCount and parentIntervalInMs, are the parameters of the resource's global statistic
+// If compliance passes, return nil, if not returns specific error
+func CheckValidityForReuseStatistic(sampleCount, intervalInMs uint32, parentSampleCount, parentIntervalInMs uint32) error {
+	if intervalInMs <= 0 || sampleCount <= 0 || intervalInMs%sampleCount != 0 {
```
Comment (yunfeiyanggzq) [inline core/base/stat.go:77]: `intervalInMs%sampleCount` I think you'd better format the code

File: core/flow/rule_manager.go (hunk 1 above)
Comment (louyuting) [reply core/flow/rule_manager.go:82]: Updated

File: core/flow/rule_manager.go (hunk 2 above)
Comment (louyuting) [reply core/flow/rule_manager.go:84]: Updated

File: core/flow/rule_manager.go (hunk 3 above)
Comment (louyuting) [reply core/flow/rule_manager.go:161]: Any suggestion?

File: core/flow/rule_manager.go (hunk 4 above)
Comment (louyuting) [reply core/flow/rule_manager.go:304]: Maybe we could improve here in future practical scenarios.
Currently, I think it's ok.

File: core/base/stat.go (hunk 5 above)
Comment (louyuting) [reply core/base/stat.go:77]: The current is the result after format(go fmt).

File: core/base/stat.go [hunk 6]
```diff
@@ -46,4 +50,47 @@ type StatNode interface {
...
+}
+
+var (
+	IllegalGlobalStatisticParamsError = errors.New("Invalid parameters, sampleCount or interval, for resource's global statistic")
+	IllegalStatisticParamsError       = errors.New("Invalid parameters, sampleCount or interval, for metric statistic")
+	GlobalStatisticNonReusableError   = errors.New("The parameters, sampleCount and interval, mismatch for reusing between resource's global statistic and readonly metric statistic.")
+)
+
+func CheckValidityForStatistic(sampleCount, intervalInMs uint32) error {
+	if intervalInMs <= 0 || sampleCount <= 0 || intervalInMs%sampleCount != 0 {
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/base/stat.go:66]: yes, that is.

File: core/config/entity.go [hunk 7]
```diff
@@ -117,6 +131,10 @@ func checkConfValid(conf *SentinelConfig) error {
 	if conf.Stat.System.CollectIntervalMs == 0 {
 		return errors.New("Bad system stat globalCfg: collectIntervalMs = 0")
 	}
+	if err := base.CheckValidityForReuseStatistic(conf.Stat.MetricStatisticSampleCount, conf.Stat.MetricStatisticIntervalMs,
+		conf.Stat.GlobalStatisticSampleCountTotal, conf.Stat.GlobalStatisticIntervalMsTotal); err != nil {
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/config/entity.go:135]: yes, conf.stat is a struct `StatConfig` rather than `*StatConfig `

File: core/flow/rule_manager.go [hunk 8]
```diff
@@ -242,13 +266,13 @@ func buildFlowMap(rules []*Rule) TrafficControllerMap {
 			tokenCalculateStrategy: rule.TokenCalculateStrategy,
 			controlBehavior:        rule.ControlBehavior,
 		}]
-		if !supported {
+		if !supported || generator == nil {
```
Comment (louyuting) [inline core/flow/rule_manager.go:269]: need to think about statistic reuse

Comment (louyuting) [reply core/flow/rule_manager.go:401]: Updated

File: core/flow/rule_manager.go [hunk 9]
```diff
@@ -270,30 +431,35 @@ func IsValidRule(rule *Rule) error {
...
+			return errors.New("invalid WarmUpPeriodSec")
 		}
 		if rule.WarmUpColdFactor == 1 {
 			return errors.New("WarmUpColdFactor must be great than 1")
 		}
 	}
+	if rule.ControlBehavior == Throttling && rule.MaxQueueingTimeMs == 0 {
+		return errors.New("invalid MaxQueueingTimeMs")
+	}
+	if !((rule.StatIntervalInMs == 0 && rule.StatSampleCount == 0) ||
```
Comment (sczyh30) [inline core/flow/rule_manager.go:460]: Maybe we could split the large condition to several checkings and make the message clearer.

File: core/flow/rule_manager.go [hunk 10]
```diff
@@ -83,8 +139,19 @@ func onRuleUpdate(rules []*Rule) (err error) {
 		}
 	}()
 
-	m := buildFlowMap(rules)
-
+	resRulesMap := make(map[string][]*Rule)
+	for _, rule := range rules {
+		if err := IsValidRule(rule); err != nil {
+			logging.Error(err, "ignoring invalid flow rule", "rule", rule)
```
Comment (sczyh30) [inline core/flow/rule_manager.go:145]: Actually the only thing we needed is error message (not error itself and its call stack), so maybe we could just extract the error message here (using warn level seems okay).

**Comment by sczyh30** [comment]
And we may need to support the scenario when only one of sampleCount or intervalMs is provided (use the default value for another).

File: core/flow/rule_manager.go (hunk 10 above)
Comment (louyuting) [reply core/flow/rule_manager.go:145]: I will update here

File: core/flow/tc_default.go [hunk 11]
```diff
@@ -5,30 +5,47 @@ import (
...
 
-func (d *DefaultTrafficShapingChecker) DoCheck(node base.StatNode, acquireCount uint32, threshold float64) *base.TokenResult {
-	if node == nil {
+func (d *RejectTrafficShapingChecker) DoCheck(resStat base.StatNode, acquireCount uint32, threshold float64) *base.TokenResult {
+	metricReadonlyStat := d.BoundOwner().boundStat.readOnlyMetric
+	if metricReadonlyStat == nil {
 		return nil
 	}
-	curCount := node.GetQPS(base.MetricEventPass)
+	curCount := metricReadonlyStat.GetQPS(base.MetricEventPass)
```
Comment (sczyh30) [inline core/flow/tc_default.go:48]: Here we may need `GetSum` instead of `GetQPS` (i.e. avg per second).

**Review by sczyh30 — CHANGES_REQUESTED** [review]

File: core/flow/tc_default.go (hunk 11 above)
Comment (louyuting) [reply core/flow/tc_default.go]: Yes, I will update here.

**Review by sczyh30 — APPROVED** [review]
LGTM

**Comment by sczyh30** [comment]
Awesome. Thanks!

**Pull request merged**